File: rgctl-tests/ecommerce-typescript/src/coolstore/models/shoppingCart.ts
import type { ShoppingCartItem } from './shoppingCartItem';

/** CoolStore-shaped cart with mutable pricing totals (CPG field-write target). */
export interface ShoppingCart {
  cartId: string;
  cartItemTotal: number;
  cartItemPromoSavings: number;
  shippingTotal: number;
  shippingPromoSavings: number;
  cartTotal: number;
  shoppingCartItemList: ShoppingCartItem[];
}

export function createShoppingCart(cartId: string): ShoppingCart {
  return {
    cartId,
    cartItemTotal: 0,
    cartItemPromoSavings: 0,
    shippingTotal: 0,
    shippingPromoSavings: 0,
    cartTotal: 0,
    shoppingCartItemList: [],
  };
}

export function resetShoppingCartItemList(cart: ShoppingCart): void {
  cart.shoppingCartItemList = [];
}

export function addShoppingCartItem(cart: ShoppingCart, sci: ShoppingCartItem): void {
  cart.shoppingCartItemList.push(sci);
}

export function removeShoppingCartItem(cart: ShoppingCart, sci: ShoppingCartItem): boolean {
  const idx = cart.shoppingCartItemList.indexOf(sci);
  if (idx >= 0) {
    cart.shoppingCartItemList.splice(idx, 1);
    return true;
  }
  return false;
}

File: rgctl-tests/ecommerce-rust/src/coolstore/model/shopping_cart.rs
use serde::{Deserialize, Serialize};

use super::ShoppingCartItem;

/// CoolStore-shaped cart with mutable pricing totals (CPG field-write target).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ShoppingCart {
    pub cart_id: String,
    pub cart_item_total: f64,
    pub cart_item_promo_savings: f64,
    pub shipping_total: f64,
    pub shipping_promo_savings: f64,
    pub cart_total: f64,
    pub shopping_cart_item_list: Vec<ShoppingCartItem>,
}

impl ShoppingCart {
    pub fn new(cart_id: impl Into<String>) -> Self {
        Self {
            cart_id: cart_id.into(),
            cart_item_total: 0.0,
            cart_item_promo_savings: 0.0,
            shipping_total: 0.0,
            shipping_promo_savings: 0.0,
            cart_total: 0.0,
            shopping_cart_item_list: Vec::new(),
        }
    }

    pub fn reset_shopping_cart_item_list(&mut self) {
        self.shopping_cart_item_list = Vec::new();
    }

    pub fn add_shopping_cart_item(&mut self, sci: ShoppingCartItem) {
        self.shopping_cart_item_list.push(sci);
    }

    pub fn remove_shopping_cart_item(&mut self, index: usize) {
        if index < self.shopping_cart_item_list.len() {
            self.shopping_cart_item_list.remove(index);
        }
    }
}
